sdks/python/src/opik/integrations/harbor/experiment_service.py

```diff
@@ -0,0 +1,206 @@
+Experiment service for Harbor integration.
+
+Manages Opik datasets and experiments for Harbor jobs, linking trial traces
+to experiments for evaluation tracking.
+"""
+
+import logging
+from datetime import datetime
+from typing import Any, Dict, Optional, Set, TYPE_CHECKING
+
+from opik.api_objects import opik_client
+from opik.api_objects.experiment import experiment_item
+
+if TYPE_CHECKING:
+    from opik.api_objects.experiment.experiment import Experiment
+    from opik.api_objects.dataset.dataset import Dataset
+
+LOGGER = logging.getLogger(__name__)
+
+# Global service instance
+_SERVICE: Optional["HarborExperimentService"] = None
+
+
+class HarborExperimentService:
+    """
+    Manages Opik datasets and experiments for Harbor jobs.
+
+    This service:
+    - Creates/gets a dataset for each Harbor dataset source
+    - Creates an experiment for the job run
+    - Links each trial's trace to the experiment as an experiment item
+    """
+
+    def __init__(
+        self,
+        experiment_name: str,
+    ) -> None:
+        """Initialize the experiment service.
+
+        Args:
+            experiment_name: Name for the experiment.
+        """
+        self._experiment_name = experiment_name
+        self._client = opik_client.get_client_cached()
+
+        # Dataset and experiment instances (created lazily per source)
+        self._datasets: Dict[str, "Dataset"] = {}
+        self._experiments: Dict[str, "Experiment"] = {}
+        self._linked_trials: Set[str] = set()
+
+    def _ensure_dataset_and_experiment(
+        self,
+        source: str,
+        job_name: Optional[str] = None,
+    ) -> None:
+        """Ensure dataset and experiment exist for a source."""
+        if source in self._experiments:
+            return
+
+        try:
+            dataset = self._client.get_or_create_dataset(
+                name=source,
+                description=f"Harbor benchmark dataset: {source}",
+            )
+            self._datasets[source] = dataset
+            LOGGER.info("Using dataset '%s' for Harbor source", source)
+
+            experiment_config: Dict[str, Any] = {"created_from": "harbor"}
+            if job_name:
+                experiment_config["job_name"] = job_name
+
+            experiment = self._client.create_experiment(
+                dataset_name=source,
+                name=self._experiment_name,
+                experiment_config=experiment_config,
+            )
+            self._experiments[source] = experiment
+            LOGGER.info(
+                "Created experiment '%s' for dataset '%s'",
+                self._experiment_name,
+                source,
+            )
+        except Exception as e:
+            LOGGER.warning(
+                "Failed to create dataset/experiment for source '%s': %s",
+                source,
+                e,
+            )
+
+    def link_trial(
+        self,
+        trial_name: str,
+        trace_id: str,
+        source: Optional[str] = None,
```

Comment:
Let's remove default values, it's not a public API, there are always values passed from another module.